lib/semantics/type.h

```diff
@@ -361,57 +359,87 @@ class DataComponentDef {
   // TODO: coarray-spec
   // TODO: component-initialization
   DataComponentDef(
-      const DeclTypeSpec &type, const Name &name, const Attrs &attrs)
+      const DeclTypeSpec &type, const SourceName &name, const Attrs &attrs)
     : DataComponentDef(type, name, attrs, ArraySpec{}) {}
-  DataComponentDef(const DeclTypeSpec &type, const Name &name,
+  DataComponentDef(const DeclTypeSpec &type, const SourceName &name,
       const Attrs &attrs, const ArraySpec &arraySpec);
 
   const DeclTypeSpec &type() const { return type_; }
-  const Name &name() const { return name_; }
+  const SourceName &name() const { return name_; }
   const Attrs &attrs() const { return attrs_; }
   const ArraySpec &shape() const { return arraySpec_; }
 
 private:
   const DeclTypeSpec type_;
-  const Name name_;
+  const SourceName name_;
   const Attrs attrs_;
   const ArraySpec arraySpec_;
   friend std::ostream &operator<<(std::ostream &, const DataComponentDef &);
 };
 
+class Symbol;
+
+// This represents a proc-interface in the declaration of a procedure or
+// procedure component. It comprises a symbol (representing the specific
+// interface), a decl-type-spec (representing the function return type),
+// or neither.
+class ProcInterface {
+public:
+  ProcInterface() = default;
+  ProcInterface(ProcInterface &&that)
+    : symbol_{that.symbol_}, type_{std::move(that.type_)} {}
+  ProcInterface(const ProcInterface &that) : symbol_{that.symbol_} {
+    if (that.type_) {
+      *this = *that.type_;
```

Comment:
Should this assignment statement be "type_ = *that.type_;"?  I don't understand what's going on otherwise.

Reply:
It's calling the operator= for the class so the logic there doesn't have to be duplicated. But it is confusing, so I'll see if I can find a clearer way to do it.

Reply:
Got it.  A short comment on the same line would be more than enough -- I always read `operator=` as setting the whole left-hand side, and didn't consider that possibility.